internal/preflight/preflight.go

```diff
@@ -63,39 +65,60 @@ func (p *preflight) PreflightUpgradeCheck(ctx context.Context, pv *kmmv1beta1.Pr
 		return false, fmt.Sprintf("Failed to substitute template in kernel mapping in the module %s for kernel version %s", mod.Name, kernelVersion)
 	}
 
+	shouldBuild := module.ShouldBeBuilt(mod.Spec, *mapping)
+	shouldSign := module.ShouldBeSigned(mod.Spec, *mapping)
+
 	err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageImage)
 	if err != nil {
 		log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to image stage"), "module", mod.Name, "error", err)
 	}
 
-	imageVerified, msg := p.helper.verifyImage(ctx, mapping, mod, kernelVersion)
-	if imageVerified || (mapping.Build == nil && mod.Spec.ModuleLoader.Container.Build == nil) {
-		return imageVerified, msg
+	verified, msg := p.helper.verifyImage(ctx, mapping, mod, kernelVersion)
+	if verified {
+		shouldBuild = false
+		shouldSign = false
 	}
 
-	err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageBuild)
-	if err != nil {
-		log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to build stage"), "module", mod.Name, "error", err)
+	if shouldBuild {
+		err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageBuild)
+		if err != nil {
+			log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to build stage"), "module", mod.Name, "error", err)
+		}
+
+		verified, msg = p.helper.verifyBuild(ctx, pv, mapping, mod)
+		if !verified {
+			shouldSign = false
+		}
 	}
 
-	return p.helper.verifyBuild(ctx, pv, mapping, mod)
+	if shouldSign {
+		err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageSign)
+		if err != nil {
```

Comment:
```suggestion
		if err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageSign); err != nil {
```

Reply:
i prefer to not change the code: i personally think it is more readable